A43-sdk-authorization.md

```diff
@@ -0,0 +1,364 @@
+service owners and/or security admins to enable per-RPC authorization checks.
+
+SDK authorization policy guarantees stability and backward compatibility. So if
+a user writes a config, it will always be supported by gRPC unless the policy
+undergoes major versioning change like adding required fields. The policy is 
+extendable, we may add new fields in the future as new requirements are introduced, 
+we won’t be removing fields or updating semantics of existing fields.
+
+Following is the JSON schema of SDK Authorization Policy Version 1.0
+
+```json
+{
+	"title": "AuthorizationPolicy",
+	"definitions": {
+		"rule": {
+			"description": "Specification of rules. An empty rule is always matched (i.e., both source and request are empty)",
+			"type" : "object",
+			"properties": {
+				"name": {
+					"description": "The name of an authorization rule. It is mainly for monitoring and error message generation.",
+					"type": "string"
+				},
+				"source": {
+					"description": "Specifies attributes of a peer. Fields in the source are ANDed together, once we support" 
+					  "multiple fields in the future. If not set, no checks will be performed against the source.",
+					"type": "object",
+					"properties": {
+						"principals": {
+							"description": "A list of peer identities to match for authorization. The principals are one"
+							  "of, i.e., it matches if one of the principals matches."
+							  "The field supports Exact, Prefix, Suffix, and Presence matches."
+							    "- Exact match: \"abc\" will match on value \"abc\"."
+							    "- Prefix match: \"abc*\" will match on value \"abc\" and \"abcd\"."
+							    "- Suffix match: \"*abc\" will match on value \"abc\" and \"xabc\"."
+							    "- Presence match: \"*\" will match when the value is not empty.",
+							"type": "array",
+							"items": {
+								"type": "string"
+							}
+						}
+					}
+				},
+				"request": {
+					"description": "Specifies attributes of a request. Fields in the request are ANDed together. If not set, no"
+					  "checks will be performed against the request.",
+					"type": "object",
+					"properties": {
+						"paths": {
+							"description": "A list of paths to match for authorization. This is the fully qualified name"
+							  "in the form of \"/package.service/method\". The paths are ORed together, i.e., it"
+							  "matches if one of the paths matches."
+							  "This field supports Exact, Prefix, Suffix, and Presence matches."
+								"- Exact match: \"abc\" will match on value \"abc\"."
+								"- Prefix match: \"abc*\" will match on value \"abc\" and \"abcd\"."
+								"- Suffix match: \"*abc\" will match on value \"abc\" and \"xabc\"."
+								"- Presence match: \"*\" will match when the value is not empty.",
+							"type": "array",
+							"items": {
+								"type": "string"
+							}
+						},
+						"headers": {
+							"description": "A list of HTTP header key/value pairs to match against, for potentially" 
```

Comment:
Is it correct that for HTTP headers, we should match them against the hard-coded values defined in [here](https://github.com/grpc/grpc/blob/master/doc/PROTOCOL-HTTP2.md)?

Reply:
Yes! Since internally in the RBAC engine we would be fetching the header from gRPC Metadata. So if ":method" is unavailable it would be hard coded to POST, and server would then deny requests for all other method types. xDS RBAC support proposal goes through the caveats in detail.

Its worth noting that we don't support all the headers in SDK API.  The following headers are *not*
supported: "hop-by-hop" headers (e.g., those listed in "Connection" header), HTTP/2 pseudo headers (":"-prefixed), the "Host" header, and headers prefixed with "grpc-". If Authorization policy has any of these headers, then the policy will be marked invalid during translation.